src/Uuid/Uuid.php

```diff
@@ -0,0 +1,274 @@
+    {
+        return '00000000-0000-0000-0000-000000000000' === $uuid;
+    }
+
+    public static function uuid_type(string $uuid)
+    {
+        if (null === $parsed = self::uuid_parse_as_array($uuid)) {
+            return false;
+        }
+
+        if (self::uuid_is_null($uuid)) {
+            return self::UUID_TYPE_NULL;
+        }
+
+        return ($parsed['time_hi_and_version'] >> 12) & 0xF;
+    }
+
+    public static function uuid_variant(string $uuid)
+    {
+        if (null === $parsed = self::uuid_parse_as_array($uuid)) {
+            return false;
+        }
+
+        if (self::uuid_is_null($uuid)) {
+            return self::UUID_TYPE_NULL;
+        }
+
+        if (($parsed['clock_seq'] & 0x8000) == 0) {
+            return self::UUID_VARIANT_NCS;
+        }
+        if (($parsed['clock_seq'] & 0x4000) == 0) {
+            return self::UUID_VARIANT_DCE;
+        }
+        if (($parsed['clock_seq'] & 0x2000) == 0) {
+            return self::UUID_VARIANT_MICROSOFT;
+        }
+
+        return self::UUID_VARIANT_OTHER;
+    }
+
+    public static function uuid_time(string $uuid)
+    {
+        if (PHP_INT_SIZE !== 8) {
+            throw new \RuntimeException('UUID time generation is not supported on 32-bit systems. Use the uuid extension instead.');
+        }
+
+        if (null === $parsed = self::uuid_parse_as_array($uuid)) {
+            return false;
+        }
+
+        if (self::UUID_TYPE_TIME !== self::uuid_type($uuid)) {
+            return false;
+        }
+
+        $high = $parsed['time_mid'] | (($parsed['time_hi_and_version'] & 0xFFF) << 16);
+        $clockReg = $parsed['time_low'] | ($high << 32);
+        $clockReg -= 122192928000000000;
+
+        return (int) ($clockReg / 10000000);
+    }
+
+    public static function uuid_mac(string $uuid)
+    {
+        if (null === $parsed = self::uuid_parse_as_array($uuid)) {
+            return false;
+        }
+
+        if (self::UUID_TYPE_TIME !== self::uuid_type($uuid)) {
+            return false;
+        }
+
+        return dechex($parsed['node']);
+    }
+
+    public static function uuid_parse(string $uuid)
+    {
+        if (null === $parsed = self::uuid_parse_as_array($uuid)) {
+            return false;
+        }
+
+        $uuid = str_replace('-', '', $uuid);
+
+        return pack('H*', $uuid);
+    }
+
+    public static function uuid_unparse(string $uuidAsBinary)
+    {
+        $data = unpack('H*', $uuidAsBinary)[1];
+
+        if (32 != \strlen($data)) {
+            return false;
+        }
+
+        $uuid = sprintf('%s-%s-%s-%s-%s',
+            substr($data, 0, 8),
+            substr($data, 8, 4),
+            substr($data, 12, 4),
+            substr($data, 16, 4),
+            substr($data, 20, 12)
+        );
+
+        if (null === self::uuid_parse_as_array($uuid)) {
+            return false;
+        }
+
+        return $uuid;
+    }
+
+    private static function uuid_generate_random(): string
```

Comment:
isn't it possible for this implementation to generate a null UUID with a non-zero probability ? And is it the same for the extension ?

Reply:
a null UUID is still a string. Did I miss something?

Reply:
well, the question is whether `uuid_generate_random()` is allowed to return this null UUID in the original implementation, or whether it never returns it (given that a null UUID won't be of type `UUID_TYPE_RANDOM` AFAICT)

Reply:
It looks like nothing [prevent](https://github.com/karelzak/util-linux/blob/6a735c2c5bfd64b06fd210663c1ebfeac6eef061/libuuid/src/gen_uuid.c#L502-L523) the extension to return a null UUID

**BUT** This is not possible since we "pad" some value for ensure the type:
```
            // 16 bits for "time_hi_and_version",
            // four most significant bits holds version number 4
            random_int(0, 0x0fff) | 0x4000,
```